### PR TITLE
Init replication

### PR DESCRIPTION
**Pull request by paultranvan**
At initialization time, we used to replicate docs from CouchDB to
PouchDB by using the replication protocol. This is problematic
because it replicates the revision history, making the indexing
potentially slower and increasing the database size.
Moreover, the replication protocol is also slower than just queyring
the _all_docs and inserting the documents in bulk.
Here are some numbers we measured on a database with 10K docs and
about 100K updates:

* Replication with CouchDB replication protocol: 34s
  * Includes revision history
* Replication with _all_docs + bulk insert: 13s
  * Does not include revision history
* Indexing with revisions: 38s
* Indexing without revisions: 26s

Note this only concern the initial replication, we continue to use the
CouchDB replication protocol for the next runs.

### CODE REVIEW

File: packages/cozy-pouch-link/src/localStorage.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  const item = window.localStorage.getItem(LOCALSTORAGE_WARMUPEDQUERIES_KEY)
+  if (!item) {
+    return {}
+  }
+  return JSON.parse(item)
+}
+
+export const destroyWarmedUpQueries = () => {
+  window.localStorage.removeItem(LOCALSTORAGE_WARMUPEDQUERIES_KEY)
+}
```
Comment (Crash--) [inline packages/cozy-pouch-link/src/localStorage.js:89]: this commit is not only about refactor. It's also about adding tests for the previous commit. Can it be split?

File: packages/cozy-pouch-link/src/startReplication.js [hunk 2]
```diff
@@ -40,9 +44,11 @@ export const startReplication = (
   replicationOptions,
   getReplicationURL
 ) => {
+  console.debug('start replication begin')
```
Comment (Crash--) [inline packages/cozy-pouch-link/src/startReplication.js:47]: to remove

File: packages/cozy-pouch-link/src/startReplication.js [hunk 3]
```diff
@@ -56,19 +62,21 @@ export const startReplication = (
...
-      // performances issues
-      const docs = await replicateAllDocs(pouch, url)
-      return resolve(docs)
+      ;(async () => {
+        // For the first remote->local replication, we manually replicate all docs
+        // as it avoids to replicate all revs history, which can lead to
+        // performances issues
+        docs = await replicateAllDocs(pouch, url)
+        return resolve(docs)
+      })()
```
Comment (Crash--) [inline packages/cozy-pouch-link/src/startReplication.js:71]: nit: I'm a big fan of having "working" commit. So if the previous solution didn't work, this change should be squashed with the "first commit"

**Review by Crash-- — COMMENTED** [review]
Looks good overall! 👍

File: packages/cozy-pouch-link/src/PouchManager.js [hunk 4]
```diff
@@ -33,13 +37,15 @@ class PouchManager {
...
     const pouchOptions = get(options, 'pouch.options', {})
+
     forEach(pouchPlugins, plugin => PouchDB.plugin(plugin))
     this.pouches = fromPairs(
       doctypes.map(doctype => [
         doctype,
         new PouchDB(this.getDatabaseName(doctype), pouchOptions)
       ])
     )
+    window.pouch = this.pouches
```
[the remark this answers is not in the document]
Comment (paultranvan) [reply packages/cozy-pouch-link/src/PouchManager.js:44]: Wops, my bad

File: packages/cozy-pouch-link/src/remote.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Fetch remote instance
+ *
+ * @param {URL} url - The remote instance URL, including the credentials
+ * @param {object} params - The params to query the remote instance
+ * @returns {object} The instance response
```
[the remark this answers is not in the document]
Comment (paultranvan) [reply packages/cozy-pouch-link/src/remote.js:6]: Not necessarily, you can have any response from the stack, typically `/_changes` will not return Documents, so it actually could be any JSON

File: packages/cozy-pouch-link/src/PouchManager.js [hunk 6]
```diff
@@ -270,6 +303,35 @@ class PouchManager {
     window.localStorage.removeItem(LOCALSTORAGE_SYNCED_KEY)
   }
 
+  persistDoctypeLastSequence(doctype, sequence) {
```
[the remark this answers is not in the document]
Comment (paultranvan) [reply packages/cozy-pouch-link/src/PouchManager.js]: You mean jsDoc for all localStorage.js methods?

File: packages/cozy-pouch-link/src/startReplication.js [hunk 7]
```diff
@@ -75,7 +85,7 @@ export const startReplication = (
           })
       }
     })
-    replication.on('error', reject).on('complete', () => {
+    replication.on('error', reject).on('complete', infos => {
```
[the remark this answers is not in the document]
Comment (paultranvan) [reply packages/cozy-pouch-link/src/startReplication.js]: Not anymore ;)

File: packages/cozy-pouch-link/src/startReplication.js [hunk 8]
```diff
@@ -97,3 +107,59 @@ export const startReplication = (
...
   return promise
 }
+
+const filterDocs = docs => {
+  return docs
+    .map(doc => doc.doc)
+    .filter(doc => !doc._deleted && !startsWith(doc._id, '_design'))
+}
+
+export const replicateAllDocs = async (db, baseUrl) => {
```
[the remark this answers is not in the document]
Comment (paultranvan) [reply packages/cozy-pouch-link/src/startReplication.js:143]: It was added in d8714785f3a3885810bb217b47cac9dfca580e93

File: packages/cozy-pouch-link/src/PouchManager.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Crash--) [reply packages/cozy-pouch-link/src/PouchManager.js]: at least for the ones you add ;)

File: packages/cozy-pouch-link/src/remote.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  Object.keys(params).forEach(key => {
+    url.searchParams.set(key, params[key])
+  })
+  const fetchUrl = `${url.protocol}//${url.href.substring(
+    url.href.indexOf('@') + 1
+  )}`
+  const headers = new Headers()
+  headers.append('Accept', 'application/json')
+  headers.append('Content-Type', 'application/json')
+  headers.append('Authorization', 'Basic ' + btoa(username + ':' + password))
```
Comment (Crash--) [inline packages/cozy-pouch-link/src/remote.js:21]: @paultranvan can't we use some methods from here: https://github.com/cozy/cozy-client/blob/master/packages/cozy-pouch-link/src/CozyPouchLink.js#L31-L35

Comment (paultranvan) [reply packages/cozy-pouch-link/src/remote.js]: Sure, f4b7e2f1bac3378f2eee367ec4ed2c4cd2fe3eaf

**Review by Crash-- — APPROVED** [review]

**Pull request merged**